contracts/Manager.sol

```diff
@@ -0,0 +1,413 @@
+        return _quantitiesRequiredToIssue(_rsvQuantity);
+    }
+
+
+    // ============================= Internals ================================
+
+    /// Internal function for all issuances to go through.
+    function _issue(uint256 _rsvQuantity) internal {
+        require(_rsvQuantity > 0, "cannot issue zero RSV");
+        uint256[] memory quantities = _quantitiesRequiredToIssue(_rsvQuantity);
+
+        // Intake collateral tokens.
+        for (uint i = 0; i < basket.size(); i++) {
+            IERC20(basket.tokens(i)).safeTransferFrom(_msgSender(), address(vault), quantities[i]);
+        }
+
+        // Compensate with RSV.
+        rsv.mint(_msgSender(), _rsvQuantity);
+
+        _assertFullyCollateralized();
+        emit Issuance(_msgSender(), _rsvQuantity);
+    }
+
+    /// Internal function for all redemptions to go through.
+    function _redeem(uint256 _rsvQuantity) internal {
+        require(_rsvQuantity > 0, "cannot redeem 0 RSV");
+
+        // Burn RSV tokens.
+        rsv.burnFrom(_msgSender(), _rsvQuantity);
+
+        // Compensate with collateral tokens.
+        vault.batchWithdrawTo(
+            basket.getTokens(), 
+            basket.quantitiesRequired(_rsvQuantity), 
+            _msgSender()
+        );
+
+        _assertFullyCollateralized();
+        emit Redemption(_msgSender(), _rsvQuantity);
+    }
+
+    /// Calculates the quantities of tokens required to issue `_rsvQuantity`. 
+    function _quantitiesRequiredToIssue(uint256 _rsvQuantity) internal view returns(uint256[] memory) {
+        uint256[] memory quantities = basket.quantitiesRequired(_rsvQuantity);
+        uint256 seigniorageMultiplier = uint256(seigniorage.add(BPS_FACTOR));
+
+        for (uint i = 0; i < basket.size(); i++) {
+            quantities[i] = quantities[i].mul(seigniorageMultiplier).div(BPS_FACTOR);
+        }
+    }
+
+    // /// Calculates the maximum we could issue to an address based on their allowances.
+    // function _calculateMaxIssuable(address funder) internal view returns(uint256) {
+    //     uint256 rsvDecimalsFactor = uint256(10) ** rsvDecimals;
+    //     uint256 allowance;
+    //     uint256 balance;
+    //     uint256 available;
+    //     uint256 issuable;
+    //     uint256 minIssuable;
+
+    //     for (uint i = 0; i < basket.size(); i ++) {
+    //         allowance = IERC20(basket.tokens(i)).allowance(funder, address(this));
+    //         balance = IERC20(basket.tokens(i)).balanceOf(funder);
+    //         available = allowance;
+    //         if (balance < available) available = balance;
+
+    //         issuable = rsvDecimalsFactor.mul(available).div(basket.backing(i));
+    //         if (issuable < minIssuable) minIssuable = issuable;
+    //     }
+    //     return minIssuable;
+    // }
+
+    /// Ensure that the Vault is fully collateralized. 
+    function _assertFullyCollateralized() internal view {
```

Comment:
Could be nice to expose this publicly. In that world I might extract a public function called `isFullyCollateralized`, that returns a bool, and then have this function just be `assert(isFullyCollateralized())`.

Reply:
Great idea.

Reply:
Done!